### PR TITLE
[CALCITE-3791] HepPlanner does not clear metadata cache for the ancestors of discarded node when a transformation happens

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-3791.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/hep/HepPlanner.java [hunk 1]
```diff
@@ -888,6 +890,28 @@ private void contractVertices(
     }
   }
 
+  /**
+   * Clears metadata cache for the RelNode and its parents.
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/plan/hep/HepPlanner.java]: parents -> ancestors?

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java [hunk 2]
```diff
@@ -943,9 +943,9 @@ private static boolean isNonNegative(Double result, boolean fail) {
    * Removes cached metadata values for specified RelNode.
    *
    * @param rel RelNode whose cached metadata should be removed
+   * @return true if cache for the provided RelNode was not empty
    */
-  public static void clearCache(RelNode rel) {
-    rel.getCluster().getMetadataQuery().clearCache(rel);
+  public static boolean clearCache(RelNode rel) {
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java]: I am not sure what is the standard procedure of changing a public API.

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java]: I think void to boolean is fine.

File: core/src/main/java/org/apache/calcite/plan/hep/HepPlanner.java (hunk 1 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/plan/hep/HepPlanner.java]: Will do.

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java (hunk 2 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java]: Agree with @hsyuan . It will not break anything.

Comment (liyafan82) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java]: Thanks a lot for your clarification.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java]: I'm wondering how to give a valid test, e.g. even though we clear the cache that should not do, the tests still passes.

Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java]: I tried to add a test case. But it is really difficult. We can see the problem is just there.

**Pull request merged**